### PR TITLE
Debug tcn

### PR DESCRIPTION
**Pull request by yanndupis**
Hey guys,

Here is a PR for the conv1D stuff. 

It handles several edge cased in the converter that where missing. The Dense layer supports transpose for the various inputs. Expand dim has been fixed. It wasn't grabbing correctly the axis input on the graph. We also decided not making the zero tensor explicit. 

Thank you!

### CODE REVIEW

File: tf_encrypted/convert/register.py [hunk 1]
```diff
@@ -125,6 +136,12 @@ def sigmoid(converter: Converter, node: Any, inputs: List[str]) -> Any:
 
 def strided_slice(converter: Converter, node: Any, inputs: List[str]) -> Any:
     input = converter.outputs[inputs[0]]
+
+    if isinstance(input, tf.NodeDef):
```
Comment (mortendahl) [inline tf_encrypted/convert/register.py]: what can `input` be here if not `NodeDef`?

File: tf_encrypted/convert/register.py [hunk 2]
```diff
@@ -363,6 +392,7 @@ def avgpool(converter: Converter, node: Any, inputs: List[str]) -> Any:
 
 
 def concat(converter: Converter, node: Any, inputs: List[str]) -> Any:
+    print("input list", inputs)
```
Comment (mortendahl) [inline tf_encrypted/convert/register.py]: leftover?

File: tf_encrypted/convert/register.py [hunk 3]
```diff
@@ -441,6 +471,8 @@ def nodef_to_public_pond(converter, x):
             nums = x.attr["value"].tensor.float_val
         elif dtype == tf.float64:
             nums = x.attr["value"].tensor.float_val
+        elif dtype == tf.int32:
```
Comment (mortendahl) [inline tf_encrypted/convert/register.py]: should `int32` be supported above as well?

File: tf_encrypted/layers/dense.py [hunk 4]
```diff
@@ -37,25 +40,39 @@ def initialize(
...
 
         self.weights = self.prot.define_private_variable(initial_weights)
-        self.bias = self.prot.define_private_variable(initial_bias)
 
     def forward(self, x):
         self.layer_input = x
-        y = x.matmul(self.weights) + self.bias
+
+        if self.transpose_input:
+            self.layer_input = self.layer_input.transpose()
```
Comment (mortendahl) [inline tf_encrypted/layers/dense.py]: would this transpose more than once if `forward` is called repeatedly?

File: tf_encrypted/layers/dense.py [hunk 5]
```diff
@@ -37,25 +40,39 @@ def initialize(
...
 
         d_weights = x.transpose().matmul(d_y)
-        d_bias = d_y.reduce_sum(axis=0)
-
         self.weights.assign((d_weights * learning_rate).neg() + self.weights)
-        self.bias.assign((d_bias * learning_rate).neg() + self.bias)
+
+        if self.bias:
+            d_bias = d_y.reduce_sum(axis=0)
+            self.bias.assign((d_bias * learning_rate).neg() + self.bias)
```
Comment (mortendahl) [inline tf_encrypted/layers/dense.py]: doesn't `assign` return an operation that has to be called for it to have any effect?

File: tf_encrypted/convert/register.py (hunk 1 above)
Comment (yanndupis) [reply tf_encrypted/convert/register.py]: if it's not a NodeDef it would be already a private pond tensor.

File: tf_encrypted/convert/register.py (hunk 3 above)
Comment (yanndupis) [reply tf_encrypted/convert/register.py]: not sure, I understand. did you mean `int64`?

File: tf_encrypted/layers/dense.py (hunk 4 above)
Comment (yanndupis) [reply tf_encrypted/layers/dense.py]: I think we should be good because every time we run forward, self.layer_input will be assign first to `x`, which is the forward input. However, we will have an issue for the weight matrix. I think we could move this weight transpose to the initialize phase.

File: tf_encrypted/layers/dense.py (hunk 5 above)
Comment (yanndupis) [reply tf_encrypted/layers/dense.py]: hmm not sure, it was like that when Morgan was training the logistic regression and it seemed to work.

Comment (jvmancuso) [reply tf_encrypted/layers/dense.py]: yes it does: https://github.com/mortendahl/tf-encrypted/blob/3a55ca33247dad91199669a72c3e51511f7b698c/tf_encrypted/protocol/pond.py#L614

**Comment by yanndupis** [comment]
Hey @mortendahl , I think all the comments should be addressed. Thanks for your review!

File: tf_encrypted/layers/dense.py (hunk 5 above)
Comment (jvmancuso) [reply tf_encrypted/layers/dense.py]: why wouldn't we do `self.bias -= (d_bias * learning_rate)` here?

Comment (yanndupis) [reply tf_encrypted/layers/dense.py]: Thanks for your comment @jvmancuso let me add this
